File: android/src/main/java/io/qwil/sensitiveinfo/SensitiveInfoPlugin.java
package io.qwil.sensitiveinfo;

import android.content.Context;
import android.content.SharedPreferences;
import android.hardware.fingerprint.FingerprintManager;
import android.os.Build;
import android.os.CancellationSignal;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyInfo;
import android.security.keystore.KeyProperties;
import android.util.Base64;
import android.util.Log;

import java.security.InvalidKeyException;
import java.security.KeyStore;
import java.util.HashMap;
import java.util.Map;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.IvParameterSpec;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;
import io.qwil.sensitiveinfo.util.AppConstants;
import io.qwil.sensitiveinfo.view.fragments.FingerprintAuthenticationDialogFragment;
import io.qwil.sensitiveinfo.view.fragments.FingerprintUiHelper;

public class SensitiveInfoPlugin implements MethodChannel.MethodCallHandler {

    // This must have 'AndroidKeyStore' as value. Unfortunately there is no predefined constant.
    private static final String ANDROID_KEYSTORE_PROVIDER = "AndroidKeyStore";

    // This is the default transformation used throughout this sample project.
    private static final String AES_DEFAULT_TRANSFORMATION = "AES/CBC/PKCS7Padding";

    private static final String KEY_ALIAS_AES = "MyAesKeyAlias";
    private static final String DELIMITER = "]";

    private FingerprintManager mFingerprintManager;
    private KeyStore mKeyStore;
    private CancellationSignal mCancellationSignal;


    private PluginRegistry.Registrar registrar;

    private SensitiveInfoPlugin(PluginRegistry.Registrar registrar) {
        this.registrar = registrar;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mFingerprintManager = (FingerprintManager) registrar.context().getSystemService(Context.FINGERPRINT_SERVICE);
            initKeyStore();
        }
    }


    @SuppressWarnings("unchecked")
    @Override
    public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {
        switch (methodCall.method) {
            case "isHardwareDetected":
                isHardwareDetected(result);
                break;
            case "hasEnrolledFingerprints":
                hasEnrolledFingerprints(result);
                break;
            case "isSensorAvailable":
                isSensorAvailable(result);
                break;
            case "getItem":
                getItem((String) methodCall.argument("key"), (Map<String, Object>) methodCall.arguments, result);
                break;
            case "setItem":
                setItem((String) methodCall.argument("key"), (String) methodCall.argument("value"), (Map<String, Object>) methodCall.arguments, result);
                break;
            case "deleteItem":
                deleteItem((String) methodCall.argument("key"), (Map<String, Object>) methodCall.arguments, result);
                break;
            case "getAllItems":
                getAllItems((Map<String, Object>) methodCall.arguments, result);
                break;
            default:
                result.notImplemented();
        }
    }

    /**
     * Checks whether the device supports fingerprint authentication and if the user has
     * enrolled at least one fingerprint.
     *
     * @return true if the user has a fingerprint capable device and has enrolled
     * one or more fingerprints
     */
    private boolean hasSetupFingerprint() {
        try {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && mFingerprintManager != null && mFingerprintManager.isHardwareDetected() && mFingerprintManager.hasEnrolledFingerprints();
        } catch (SecurityException e) {
            // Should never be thrown since we have declared the USE_FINGERPRINT permission
            // in the manifest file
            return false;
        }
    }

    private void isHardwareDetected(final MethodChannel.Result pm) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            pm.success(mFingerprintManager.isHardwareDetected());
        } else {
            pm.success(false);
        }
    }

    private void hasEnrolledFingerprints(final MethodChannel.Result pm) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            pm.success(mFingerprintManager.hasEnrolledFingerprints());
        } else {
            pm.success(false);
        }
    }

    private void isSensorAvailable(final MethodChannel.Result promise) {
        if (hasSetupFingerprint()) {
            promise.success("Touch ID");
        } else {
            promise.success("none");
        }
    }

    private void getItem(String key, Map<String, Object> options, MethodChannel.Result pm) {

        String name = sharedPreferences(options);

        String value = prefs(name).getString(key, null);

        if (value != null && options.containsKey("biometric") && (Boolean) options.get("biometric")) {
            boolean showModal = options.containsKey("showModal") && (Boolean) options.get("showModal");
            //noinspection unchecked
            HashMap strings = options.containsKey("strings") ? (HashMap<String, Object>) options.get("strings") : new HashMap();

            decryptWithAes(value, showModal, strings, pm, null);
        } else {
            pm.success(value);
        }
    }

    private void setItem(String key, String value, Map<String, Object> options, MethodChannel.Result pm) {
        String name = sharedPreferences(options);

        if (options.containsKey("biometric") && (Boolean) options.get("biometric")) {
            boolean showModal = options.containsKey("showModal") && (Boolean) options.get("showModal");
            //noinspection unchecked
            HashMap strings = options.containsKey("strings") ? (HashMap<String, Object>) options.get("strings") : new HashMap();

            putExtraWithAES(key, value, prefs(name), showModal, strings, pm, null);
        } else {
            try {
                putExtra(key, value, prefs(name));
                pm.success(value);
            } catch (Exception e) {
                pm.error("SensitiveInfo", e.getMessage(), null);
            }
        }
    }

    private void deleteItem(String key, Map<String, Object> options, MethodChannel.Result pm) {

        String name = sharedPreferences(options);

        SharedPreferences.Editor editor = prefs(name).edit();

        editor.remove(key).apply();

        pm.success(null);
    }

    private void getAllItems(Map<String, Object> options, MethodChannel.Result pm) {

        String name = sharedPreferences(options);

        Map<String, ?> allEntries = prefs(name).getAll();
        HashMap<String, Object> resultData = new HashMap<>();

        for (Map.Entry<String, ?> entry : allEntries.entrySet()) {
            String value = entry.getValue().toString();
            resultData.put(entry.getKey(), value);
        }
        pm.success(resultData);
    }

    @SuppressWarnings("unused")
    public void cancelFingerprintAuth() {
        if (mCancellationSignal != null && !mCancellationSignal.isCanceled()) {
            mCancellationSignal.cancel();
        }
    }

    private SharedPreferences prefs(String name) {
        return registrar.context().getSharedPreferences(name, Context.MODE_PRIVATE);
    }


    private String sharedPreferences(Map<String, Object> options) {
        String name = options.containsKey("sharedPreferencesName") ? (String) options.get("sharedPreferencesName") : "shared_preferences";
        if (name == null) {
            name = "shared_preferences";
        }
        return name;
    }


    private void putExtra(String key, Object value, SharedPreferences mSharedPreferences) {
        SharedPreferences.Editor editor = mSharedPreferences.edit();
        if (value instanceof String) {
            editor.putString(key, (String) value).apply();
        } else if (value instanceof Boolean) {
            editor.putBoolean(key, (Boolean) value).apply();
        } else if (value instanceof Integer) {
            editor.putInt(key, (Integer) value).apply();
        } else if (value instanceof Long) {
            editor.putLong(key, (Long) value).apply();
        } else if (value instanceof Float) {
            editor.putFloat(key, (Float) value).apply();
        }
    }

    /**
     * Generates a new AES key and stores it under the { @code KEY_ALIAS_AES } in the
     * Android Keystore.
     */
    private void initKeyStore() {
        try {
            mKeyStore = KeyStore.getInstance(ANDROID_KEYSTORE_PROVIDER);
            mKeyStore.load(null);

            // Check if a generated key exists under the KEY_ALIAS_AES .
            if (!mKeyStore.containsAlias(KEY_ALIAS_AES)) {
                prepareKey();
            }
        } catch (Exception ignored) {
        }
    }

    private void prepareKey() throws Exception {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.M) {
            return;
        }
        KeyGenerator keyGenerator = KeyGenerator.getInstance(
                KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE_PROVIDER);

        KeyGenParameterSpec.Builder builder;
        builder = new KeyGenParameterSpec.Builder(
                KEY_ALIAS_AES,
                KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT);

        builder.setBlockModes(KeyProperties.BLOCK_MODE_CBC)
                .setKeySize(256)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
                // forces user authentication with fingerprint
                .setUserAuthenticationRequired(true);

        keyGenerator.init(builder.build());
        keyGenerator.generateKey();
    }

    private void putExtraWithAES(final String key, final String value, final SharedPreferences mSharedPreferences, final boolean showModal, final HashMap strings, final MethodChannel.Result pm, Cipher cipher) {

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M && hasSetupFingerprint()) {
            try {
                if (cipher == null) {
                    SecretKey secretKey = (SecretKey) mKeyStore.getKey(KEY_ALIAS_AES, null);
                    cipher = Cipher.getInstance(AES_DEFAULT_TRANSFORMATION);
                    cipher.init(Cipher.ENCRYPT_MODE, secretKey);


                    // Retrieve information about the SecretKey from the KeyStore.
                    SecretKeyFactory factory = SecretKeyFactory.getInstance(
                            secretKey.getAlgorithm(), ANDROID_KEYSTORE_PROVIDER);
                    KeyInfo info = (KeyInfo) factory.getKeySpec(secretKey, KeyInfo.class);

                    if (info.isUserAuthenticationRequired() &&
                            info.getUserAuthenticationValidityDurationSeconds() == -1) {

                        if (showModal) {

                            // define class as a callback
                            class PutExtraWithAESCallback implements FingerprintUiHelper.Callback {
                                @Override
                                public void onAuthenticated(FingerprintManager.AuthenticationResult result) {
                                    //noinspection ConstantConditions
                                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                        putExtraWithAES(key, value, mSharedPreferences, true, strings, pm, result.getCryptoObject().getCipher());
                                    }
                                }

                                @Override
                                public void onError(String errorCode, CharSequence errString) {
                                    pm.error(String.valueOf(errorCode), errString.toString(), null);
                                }
                            }

                            // Show the fingerprint dialog
                            FingerprintAuthenticationDialogFragment fragment
                                    = FingerprintAuthenticationDialogFragment.newInstance(strings);
                            fragment.setCryptoObject(new FingerprintManager.CryptoObject(cipher));
                            fragment.setCallback(new PutExtraWithAESCallback());

                            fragment.show(registrar.activity().getFragmentManager(), AppConstants.DIALOG_FRAGMENT_TAG);

                        } else {
                            mCancellationSignal = new CancellationSignal();
                            mFingerprintManager.authenticate(new FingerprintManager.CryptoObject(cipher), mCancellationSignal,
                                    0, new FingerprintManager.AuthenticationCallback() {

                                        @Override
                                        public void onAuthenticationFailed() {
                                            super.onAuthenticationFailed();
                                        }

                                        @Override
                                        public void onAuthenticationError(int errorCode, CharSequence errString) {
                                            super.onAuthenticationError(errorCode, errString);
                                            pm.error(String.valueOf(errorCode), errString.toString(), null);
                                        }

                                        @Override
                                        public void onAuthenticationHelp(int helpCode, CharSequence helpString) {
                                            super.onAuthenticationHelp(helpCode, helpString);
                                        }

                                        @Override
                                        public void onAuthenticationSucceeded(FingerprintManager.AuthenticationResult result) {
                                            super.onAuthenticationSucceeded(result);
                                            //noinspection ConstantConditions
                                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                                putExtraWithAES(key, value, mSharedPreferences, false, strings, pm, result.getCryptoObject().getCipher());
                                            }
                                        }
                                    }, null);
                        }
                    }
                    return;
                }
                byte[] encryptedBytes = cipher.doFinal(value.getBytes());

                // Encode the initialization vector (IV) and encryptedBytes to Base64.
                String base64IV = Base64.encodeToString(cipher.getIV(), Base64.DEFAULT);
                String base64Cipher = Base64.encodeToString(encryptedBytes, Base64.DEFAULT);

                String result = base64IV + DELIMITER + base64Cipher;

                putExtra(key, result, mSharedPreferences);
                pm.success(value);
            } catch (InvalidKeyException e) {
                try {
                    mKeyStore.deleteEntry(KEY_ALIAS_AES);
                    prepareKey();
                } catch (Exception keyResetError) {
                    pm.error("SensitiveInfo", keyResetError.getCause().getMessage(), null);
                }
                pm.error("SensitiveInfo", e.getMessage(), null);
            } catch (SecurityException e) {
                pm.error("SensitiveInfo", e.getMessage(), null);
            } catch (Exception e) {
                pm.error("SensitiveInfo", e.getMessage(), null);
            }
        } else {
            pm.error("Fingerprint not supported", "Fingerprint not supported", null);
        }
    }

    private void decryptWithAes(final String encrypted, final boolean showModal, final HashMap strings, final MethodChannel.Result pm, Cipher cipher) {

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M
                && hasSetupFingerprint()) {

            String[] inputs = encrypted.split(DELIMITER);
            if (inputs.length < 2) {
                pm.error("DecryptionFailed", "DecryptionFailed", null);
            }

            try {
                byte[] iv = Base64.decode(inputs[0], Base64.DEFAULT);
                byte[] cipherBytes = Base64.decode(inputs[1], Base64.DEFAULT);

                if (cipher == null) {
                    SecretKey secretKey = (SecretKey) mKeyStore.getKey(KEY_ALIAS_AES, null);
                    cipher = Cipher.getInstance(AES_DEFAULT_TRANSFORMATION);
                    cipher.init(Cipher.DECRYPT_MODE, secretKey, new IvParameterSpec(iv));

                    SecretKeyFactory factory = SecretKeyFactory.getInstance(
                            secretKey.getAlgorithm(), ANDROID_KEYSTORE_PROVIDER);
                    KeyInfo info = (KeyInfo) factory.getKeySpec(secretKey, KeyInfo.class);

                    if (info.isUserAuthenticationRequired() &&
                            info.getUserAuthenticationValidityDurationSeconds() == -1) {

                        if (showModal) {

                            // define class as a callback
                            class DecryptWithAesCallback implements FingerprintUiHelper.Callback {
                                @Override
                                public void onAuthenticated(FingerprintManager.AuthenticationResult result) {
                                    //noinspection ConstantConditions
                                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                        decryptWithAes(encrypted, true, strings, pm, result.getCryptoObject().getCipher());
                                    }
                                }

                                @Override
                                public void onError(String errorCode, CharSequence errString) {
                                    pm.error(String.valueOf(errorCode), errString.toString(), null);
                                }
                            }

                            // Show the fingerprint dialog
                            FingerprintAuthenticationDialogFragment fragment
                                    = FingerprintAuthenticationDialogFragment.newInstance(strings);
                            fragment.setCryptoObject(new FingerprintManager.CryptoObject(cipher));
                            fragment.setCallback(new DecryptWithAesCallback());

                            fragment.show(registrar.activity().getFragmentManager(), AppConstants.DIALOG_FRAGMENT_TAG);

                        } else {
                            mCancellationSignal = new CancellationSignal();
                            mFingerprintManager.authenticate(new FingerprintManager.CryptoObject(cipher), mCancellationSignal,
                                    0, new FingerprintManager.AuthenticationCallback() {

                                        @Override
                                        public void onAuthenticationFailed() {
                                            super.onAuthenticationFailed();

                                        }

                                        @Override
                                        public void onAuthenticationError(int errorCode, CharSequence errString) {
                                            super.onAuthenticationError(errorCode, errString);
                                            pm.error(String.valueOf(errorCode), errString.toString(), null);
                                        }

                                        @Override
                                        public void onAuthenticationHelp(int helpCode, CharSequence helpString) {
                                            super.onAuthenticationHelp(helpCode, helpString);
                                        }

                                        @Override
                                        public void onAuthenticationSucceeded(FingerprintManager.AuthenticationResult result) {
                                            super.onAuthenticationSucceeded(result);
                                            //noinspection ConstantConditions
                                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                                decryptWithAes(encrypted, false, strings, pm, result.getCryptoObject().getCipher());
                                            }
                                        }
                                    }, null);
                        }
                    }
                    return;
                }
                byte[] decryptedBytes = cipher.doFinal(cipherBytes);
                pm.success(new String(decryptedBytes));
            } catch (InvalidKeyException e) {
                try {
                    mKeyStore.deleteEntry(KEY_ALIAS_AES);
                    prepareKey();
                } catch (Exception keyResetError) {
                    pm.error("SensitiveInfo", keyResetError.getCause().getMessage(), null);
                }
                pm.error("SensitiveInfo", e.getMessage(), null);
            } catch (SecurityException e) {
                pm.error("SensitiveInfo", e.getMessage(), null);
            } catch (Exception e) {
                pm.error("SensitiveInfo", e.getMessage(), null);
            }
        } else {
            pm.error("Fingerprint not supported", "Fingerprint not supported", null);
        }
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(PluginRegistry.Registrar registrar) {
        final MethodChannel channel =
                new MethodChannel(registrar.messenger(), "io.qwil/sensitive_info");
        channel.setMethodCallHandler(new SensitiveInfoPlugin(registrar));
    }
}

File: android/src/main/kotlin/io/qwil/sensitiveinfo/SensitiveInfoPlugin.kt
@file:Suppress("NAME_SHADOWING")

package io.qwil.sensitiveinfo

import android.content.Context
import android.content.SharedPreferences
import android.hardware.fingerprint.FingerprintManager
import android.os.Build
import android.os.CancellationSignal
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyInfo
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import java.security.KeyStore
import java.util.*
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.IvParameterSpec

class SensitiveInfoPlugin(private val registrar: Registrar) : MethodCallHandler {

    private lateinit var mFingerprintManager: FingerprintManager
    private lateinit var mKeyStore: KeyStore
    private lateinit var mCancellationSignal: CancellationSignal

    init {
        initKeyStore()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mFingerprintManager = registrar.context().getSystemService(Context.FINGERPRINT_SERVICE) as FingerprintManager
        }
    }


    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "isSensorAvailable" -> isSensorAvailable(result)
            "getItem" -> getItem(call, result)
            "setItem" -> setItem(call, result)
            "deleteItem" -> deleteItem(call, result)
            "getAllItems" -> getAllItems(call, result)
            "cancelFingerprintAuth" -> cancelFingerprintAuth()
            else -> result.notImplemented()
        }
    }

    private fun isSensorAvailable(result: Result) = when {
        hasSetupFingerprint() -> result.success("Touch ID")
        else -> result.success("none")
    }

    private fun getItem(method: MethodCall, result: Result) {
        val name = sharedPreferences(method)
        val key = method.argument<String>("key")

        val value = prefs(name).getString(key, null)
        val biometric = method.argument<Boolean>("biometric")

        if (value != null && biometric != null && biometric) {
            decryptWithAes(value, result, null)
        } else {
            result.success(value)
        }
    }

    private fun setItem(method: MethodCall, result: Result) {
        val name = sharedPreferences(method)

        val biometric = method.argument<Boolean>("biometric")
        val key = method.argument<String>("key")
        val value = method.argument<String>("value")

        if (biometric != null && biometric) {
            putExtraWithAES(key, value, prefs(name), result, null)
        } else {
            try {
                putExtra(key, value, prefs(name))
                result.success(value)
            } catch (e: Exception) {
                Log.d(TAG, e.cause?.message)
                result.error(TAG, e.message, null)
            }

        }
    }

    private fun deleteItem(method: MethodCall, result: Result) {
        val name = sharedPreferences(method)
        val key = method.argument<String>("key")
        val editor = prefs(name).edit()
        editor.remove(key).apply()
        result.success(null)
    }

    private fun getAllItems(method: MethodCall, result: Result) {
        val name = sharedPreferences(method)

        val allEntries = prefs(name).all
        val resultData = HashMap<String, Any>()

        for (entry in allEntries.entries) {
            val value = entry.value.toString()
            resultData[entry.key] = value
        }
        result.success(resultData)
    }

    private fun cancelFingerprintAuth() {
        if (!mCancellationSignal.isCanceled) {
            mCancellationSignal.cancel()
        }
    }


    private fun hasSetupFingerprint(): Boolean {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !(!mFingerprintManager.isHardwareDetected || !mFingerprintManager.hasEnrolledFingerprints())
    }

    private fun prefs(name: String): SharedPreferences {
        return registrar.context().getSharedPreferences(name, Context.MODE_PRIVATE)
    }

    private fun sharedPreferences(methodCall: MethodCall): String {
        val sharedPreferencesName = methodCall.argument<String>("keychainName")
        return sharedPreferencesName ?: "shared_preferences"
    }


    private fun putExtra(key: String, value: Any, mSharedPreferences: SharedPreferences) {
        val editor = mSharedPreferences.edit()
        when (value) {
            is String -> editor.putString(key, value).apply()
            is Boolean -> editor.putBoolean(key, value).apply()
            is Int -> editor.putInt(key, value).apply()
            is Long -> editor.putLong(key, value).apply()
            is Float -> editor.putFloat(key, value).apply()
        }
    }

    private fun initKeyStore() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return
        try {
            mKeyStore = KeyStore.getInstance(ANDROID_KEYSTORE_PROVIDER)
            mKeyStore.load(null)

            // Check if a generated key exists under the KEY_ALIAS_AES .
            if (!mKeyStore.containsAlias(KEY_ALIAS_AES)) {
                val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE_PROVIDER)

                val builder = KeyGenParameterSpec.Builder(
                        KEY_ALIAS_AES,
                        KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)

                builder.setBlockModes(KeyProperties.BLOCK_MODE_CBC)
                        .setKeySize(256)
                        .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
                        // forces user authentication with fingerprint
                        .setUserAuthenticationRequired(true)

                keyGenerator.init(builder.build())
                keyGenerator.generateKey()
            }
        } catch (ignored: Exception) {
        }

    }

    private fun putExtraWithAES(key: String, value: String, mSharedPreferences: SharedPreferences, methodResult: Result, cipher: Cipher?) {
        if (hasSetupFingerprint()) {
            try {
                var cipher = cipher
                if (cipher == null) {
                    val secretKey = mKeyStore.getKey(KEY_ALIAS_AES, null) as SecretKey
                    cipher = Cipher.getInstance(AES_DEFAULT_TRANSFORMATION)
                    cipher!!.init(Cipher.ENCRYPT_MODE, secretKey)

                    // Retrieve information about the SecretKey from the KeyStore.
                    val factory = SecretKeyFactory.getInstance(secretKey.algorithm, ANDROID_KEYSTORE_PROVIDER)
                    val info = factory.getKeySpec(secretKey, KeyInfo::class.java) as KeyInfo

                    if (info.isUserAuthenticationRequired && info.userAuthenticationValidityDurationSeconds == -1) {
                        mCancellationSignal = CancellationSignal()
                        mFingerprintManager.authenticate(FingerprintManager.CryptoObject(cipher), mCancellationSignal, 0, object : FingerprintManager.AuthenticationCallback() {

                            override fun onAuthenticationFailed() {
                                super.onAuthenticationFailed()
                                methodResult.error("FINGERPRINT_AUTHENTICATION_HELP", "Fingerprint not recognized.", null)
                            }

                            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                                super.onAuthenticationError(errorCode, errString)
                                methodResult.error(errorCode.toString(), errString.toString(), null)
                            }

                            override fun onAuthenticationHelp(helpCode: Int, helpString: CharSequence) {
                                super.onAuthenticationHelp(helpCode, helpString)
                                methodResult.error("FINGERPRINT_AUTHENTICATION_HELP", helpString.toString(), null)
                            }

                            override fun onAuthenticationSucceeded(result: FingerprintManager.AuthenticationResult) {
                                super.onAuthenticationSucceeded(result)
                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                    putExtraWithAES(key, value, mSharedPreferences, methodResult, result.cryptoObject.cipher)
                                }
                            }
                        }, null)
                    }
                    return
                }
                val encryptedBytes = cipher.doFinal(value.toByteArray())

                // Encode the initialization vector (IV) and encryptedBytes to Base64.
                val base64IV = Base64.encodeToString(cipher.iv, Base64.DEFAULT)
                val base64Cipher = Base64.encodeToString(encryptedBytes, Base64.DEFAULT)

                val result = base64IV + DELIMITER + base64Cipher

                putExtra(key, result, mSharedPreferences)
                methodResult.success(value)
            } catch (e: SecurityException) {
                methodResult.error(null, e.message, null)
            } catch (e: Exception) {
                methodResult.error(null, e.message, null)
            }
        } else {
            methodResult.error(null, "Fingerprint not supported", null)
        }
    }

    private fun decryptWithAes(encrypted: String, methodResult: Result, cipher: Cipher?) {
        var cipher = cipher
        if (hasSetupFingerprint()) {

            val inputs = encrypted.split(DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (inputs.size < 2) {
                methodResult.error(null, "DecryptionFailed", null)
                return
            }

            try {
                val iv = Base64.decode(inputs[0], Base64.DEFAULT)
                val cipherBytes = Base64.decode(inputs[1], Base64.DEFAULT)

                if (cipher == null) {
                    val secretKey = mKeyStore.getKey(KEY_ALIAS_AES, null) as SecretKey
                    cipher = Cipher.getInstance(AES_DEFAULT_TRANSFORMATION)
                    cipher!!.init(Cipher.DECRYPT_MODE, secretKey, IvParameterSpec(iv))

                    val factory = SecretKeyFactory.getInstance(
                            secretKey.algorithm, ANDROID_KEYSTORE_PROVIDER)
                    val info = factory.getKeySpec(secretKey, KeyInfo::class.java) as KeyInfo

                    if (info.isUserAuthenticationRequired && info.userAuthenticationValidityDurationSeconds == -1) {

                        mCancellationSignal = CancellationSignal()
                        mFingerprintManager.authenticate(FingerprintManager.CryptoObject(cipher), mCancellationSignal,
                                0, object : FingerprintManager.AuthenticationCallback() {

                            override fun onAuthenticationFailed() {
                                super.onAuthenticationFailed()
                                Log.d(TAG, "onAuthenticationFailed: ")
                                methodResult.error("FINGERPRINT_AUTHENTICATION_HELP", "Fingerprint not recognized.", null)
                            }

                            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                                super.onAuthenticationError(errorCode, errString)
                                Log.d(TAG, "onAuthenticationError: $errorCode")
                                methodResult.error(errorCode.toString(), errString.toString(), null)
                            }

                            override fun onAuthenticationHelp(helpCode: Int, helpString: CharSequence) {
                                super.onAuthenticationHelp(helpCode, helpString)
                                Log.d(TAG, "onAuthenticationHelp: $helpCode")
                                methodResult.error("FINGERPRINT_AUTHENTICATION_HELP", helpString.toString(), null)
                            }

                            override fun onAuthenticationSucceeded(result: FingerprintManager.AuthenticationResult) {
                                super.onAuthenticationSucceeded(result)
                                Log.d(TAG, "onAuthenticationSucceeded: " + result.toString())

                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                    decryptWithAes(encrypted, methodResult, result.cryptoObject.cipher)
                                }
                            }
                        }, null)
                    }
                    return
                }
                val decryptedBytes = cipher.doFinal(cipherBytes)

                Log.d("FlutterSensitiveInfo", String(decryptedBytes) + " sa")

                methodResult.success(String(decryptedBytes))
            } catch (e: SecurityException) {
                methodResult.error(null, e.message, null)
            } catch (e: Exception) {
                methodResult.error(null, e.message, null)
            }

        } else {
            methodResult.error(null, "Fingerprint not supported", null)
        }
    }


    companion object {
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            val channel = MethodChannel(registrar.messenger(), "io.qwil/sensitive_info")
            channel.setMethodCallHandler(SensitiveInfoPlugin(registrar))
        }

        const val TAG = "SensitiveInfoPlugin"

        // This must have 'AndroidKeyStore' as value. Unfortunately there is no predefined constant.
        private const val ANDROID_KEYSTORE_PROVIDER = "AndroidKeyStore"

        // This is the default transformation used throughout this sample project.
        private const val AES_DEFAULT_TRANSFORMATION = "AES/CBC/PKCS7Padding"

        private const val KEY_ALIAS_AES = "MyAesKeyAlias"
        private const val DELIMITER = "]"
    }

}
